ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/functions/HiveMergeablAggregate.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.hadoop.hive.ql.optimizer.calcite.functions;
+
+import org.apache.calcite.sql.SqlAggFunction;
+import org.apache.calcite.sql.SqlFunctionCategory;
+import org.apache.calcite.sql.SqlKind;
+import org.apache.calcite.sql.type.SqlOperandTypeChecker;
+import org.apache.calcite.sql.type.SqlOperandTypeInference;
+import org.apache.calcite.sql.type.SqlReturnTypeInference;
+
+/**
+ * Mergeable aggregate.
+ *
+ * A mergeable aggregate is:
+ * - accepts the same kind as inputs as the output (an X^n -> X function)
+ *
+ * Example: the SUM function is a great example; since SUM of SUM -s is the overall sum.
+ */
+public class HiveMergeablAggregate extends SqlAggFunction  {
```

Comment:
Is this the same mechanism that `SqlSplittableAggFunction` is providing in Calcite? If it is not, how is it different?

If `SqlSplittableAggFunction` can be used, we should use existing interface.

Reply:
That interface is quite wide; we might be able to utilize it - but to do that we should also implement functions like `singleton` (calculate the agg for a single value) which is trivial for a `COUNT` or a `SUM` (the interface is only implemented for things like these right now) ; but for a more complicated thing - like a sketch function it should involve executing the actual function to fold the single value.

I think it would be better to have a set of smaller interfaces - instead of one bigger; that could make it easier to implement the most valuable ones for a given usecase; when the others are not neccessarily needed.

This could be probably extended to support the `SqlSplittableAggFunction` interface - but as you can see there will be some challenges along the way ; so if we want to do that - I think it should at least have it's own ticket.

Reply:
opened: HIVE-23092 to investigate whether the Splittable interface could be utilized

Reply:
Sure, that sounds fine. If there is a point in having a smaller interface, probably it makes to discuss it with Calcite community and make those changes over there; that way we will not end up having different interfaces in both projects targeting same use case.